os/kernel/task_monitor/task_monitor.c

```diff
@@ -19,32 +19,117 @@
+
 #include <sys/boardctl.h>
 #include <tinyara/sched.h>
+
 #include "task_monitor_internal.h"
 
-static int monitored_tasks_list[CONFIG_MAX_TASKS];
+static task_monitor_node_t g_monitored_tasks_list[CONFIG_MAX_TASKS];
+static task_monitor_node_queue_t g_que_list[TASK_MONITOR_CHECK_TIME];
+static int g_monitor_cnt;
+static sem_t g_stop_sem;
+
+int task_monitor_register_list(int pid, int interval)
+{
+	if (interval < CONFIG_TASK_MONITOR_INTERVAL || CONFIG_TASK_MONITOR_MAX_INTERVAL < interval) {
+		dbg("Must be greater than %d and less than or equal to %d.\n", CONFIG_TASK_MONITOR_INTERVAL, CONFIG_TASK_MONITOR_MAX_INTERVAL);
+		return EINVAL;
+	}
+
+	int hash_pid = PIDHASH(pid);
+	if (g_monitored_tasks_list[hash_pid].interval != 0) {
+		dbg("Already registered.\n");
+		return EALREADY;
+	}
+
+	g_monitored_tasks_list[hash_pid].pid = pid;
+	if (interval % CONFIG_TASK_MONITOR_INTERVAL == 0) {
+		g_monitored_tasks_list[hash_pid].interval = interval / CONFIG_TASK_MONITOR_INTERVAL;
+	} else {
+		/* If a value that is not divided, it is rounded up. */
+		g_monitored_tasks_list[hash_pid].interval = (interval / CONFIG_TASK_MONITOR_INTERVAL) + 1;
+		if (g_monitored_tasks_list[hash_pid].interval > TASK_MONITOR_CHECK_TIME) {
+			g_monitored_tasks_list[hash_pid].interval--;
+		}
+	}
+
+	dq_addlast((FAR dq_entry_t *)&g_monitored_tasks_list[hash_pid], &g_que_list[interval - 1].q);
+
+	g_monitor_cnt++;
+	if (g_monitor_cnt == 1) {
+		sem_post(&g_stop_sem);
+	}
+	return OK;
+}
 
-void task_monitor_update_list(int pid, int status)
+void task_monitor_unregester_list(int pid)
 {
-	monitored_tasks_list[PIDHASH(pid)] = status;
+	int hash_pid;
+	int interval;
+	hash_pid = PIDHASH(pid);
+
+	if (g_monitored_tasks_list[hash_pid].interval == 0) {
+		/* Not registered. */
+		return;
+	}
+
+	g_monitored_tasks_list[hash_pid].pid = 0;
+	interval = g_monitored_tasks_list[hash_pid].interval;
+	g_monitored_tasks_list[hash_pid].interval = 0;
+
+	dq_rem((FAR dq_entry_t *)&g_monitored_tasks_list[hash_pid], &g_que_list[interval - 1].q);
+
+	g_monitor_cnt--;
 }
 
+void task_monitor_init(void)
+{
+	int pid_idx;
+	int time_idx;
+
+	/* Initialize the dequeue list */
+	for (time_idx = 0; time_idx < TASK_MONITOR_CHECK_TIME; ++time_idx) {
+		g_que_list[time_idx].remaining_count = time_idx;
+		dq_init(&(g_que_list[time_idx].q));
+	}
+
+	/* Initialize the monitored tasks list */
+	for (pid_idx = 0; pid_idx < CONFIG_MAX_TASKS; pid_idx++) {
+		g_monitored_tasks_list[pid_idx].flink = NULL;
+		g_monitored_tasks_list[pid_idx].blink = NULL;
+		g_monitored_tasks_list[pid_idx].pid = 0;
+		g_monitored_tasks_list[pid_idx].interval = 0;
+	}
+
+	g_monitor_cnt = 0;
+	sem_init(&g_stop_sem, 0, 0);
+}
 /****************************************************************************
  * Main Function
  ****************************************************************************/
 int task_monitor(int argc, char *argv[])
 {
-	int pid_idx;
+	int time_idx;
 	struct tcb_s *tcb;
+	task_monitor_node_t *next_mon;
```

Comment:
I think `next_node` or `next_mon_node` is better, because this list has monitor_nodes, not monitor.